### PR TITLE
Remove FOR_EACH_ACTIVE_CHANNEL macro

### PR DESCRIPTION
**Pull request by peterbarker**
Sadly (and a little surprisingly), this currently makes fmuv3 build 72 bytes larger for me :-(

### CODE REVIEW

File: libraries/GCS_MAVLink/GCS.cpp [hunk 1]
```diff
@@ -20,32 +20,87 @@ void GCS::send_text(MAV_SEVERITY severity, const char *fmt, ...)
...
+void GCS::send_to_active_channels(const mavlink_message_t &msg)
+{
+    for (uint8_t i=0; i<num_gcs(); i++) {
+        if (!chan(i).initialised) {
+            continue;
+        }
+        if (!(GCS_MAVLINK::active_channel_mask() & (1 << (chan(i).get_chan()-MAVLINK_COMM_0)))) {
+            continue;
+        }
+        chan(i).send_message(msg);
```
Comment (WickedShell) [inline libraries/GCS_MAVLink/GCS.cpp]: Does this do the space checking for the message?

Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS.cpp]: Caller's responsibility.  I've added a check into this method.

I've also added another patch with a SITL sanity check for attempting to write too much into a uart.  Seems that might have picked something up.  May split that patch into a separate PR.

Comment (WickedShell) [reply libraries/GCS_MAVLink/GCS.cpp]: Because this iterates all the channels how can the caller check in advance?

**Comment by peterbarker** [comment]
On Sat, 16 Feb 2019, WickedShell wrote:

> Because this iterates all the channels how can the caller check in advance?

*this* function checks and fails silently on that channel :-)

**Comment by WickedShell** [comment]
> > On Sat, 16 Feb 2019, WickedShell wrote: Because this iterates all the channels how can the caller check in advance?

> *this* function checks and fails silently on that channel :-)

It does *now* :D

File: libraries/AP_Button/AP_Button.cpp [hunk 2]
```diff
@@ -142,21 +142,14 @@ void AP_Button::timer_update(void)
...
-        mavlink_channel_t chan = (mavlink_channel_t)i;
-        if (HAVE_PAYLOAD_SPACE(chan, BUTTON_CHANGE)) {
-            mavlink_msg_button_change_send(chan,
-                                           now,
-                                           (uint32_t)last_change_time_ms,
-                                           last_mask);
-        }
-    }
+    mavlink_message_t msg;
+    mavlink_msg_button_change_pack(mavlink_system.sysid,
```
Comment (OXINARF) [inline libraries/AP_Button/AP_Button.cpp]: Sorry to crush your dreams, but this doesn't work. You are packing the message just for channel 0, which may have different proprieties to the channel the message actually ends up being sent to.

**Comment by peterbarker** [comment]
On Fri, 22 Feb 2019, Francisco Ferreira wrote:

> This has been attempted before, unfortunately with poor results. It would be possible to have a method like this:
> 
> send_to_active_channels(uint32_t msgid, const char *packet, uint8_t min_length, uint8_t length, uint8_t crc_extra)
> 
> and then call _mav_finalize_message_chan_send. It's not pretty though.

I've been looking at doing something similar.  I think we can actually do 
better than that, but it will require some changes to mavlink so we can 
retrieve "uint8_t length" from it.  We can already get crc_extra and 
min_length ('though I need to chat with tridge about why we're supply 
min_length in a field called "length" at some stage.... and what the 
implications of a memset-to-zero which uses that are...

And yes, I think this may fix the send-to-components thing, too.

> Sorry to crush your dreams, but this doesn't work. You are packing the message just for channel 0, which may have different proprieties to the channel the
> message actually ends up being sent to.

Yes, I'm hoping to be able to finalize it for each channel.

**Comment by peterbarker** [comment]
Should not go in before https://github.com/ArduPilot/ardupilot/pull/10616

File: libraries/AP_Button/AP_Button.cpp (hunk 2 above)
Comment (peterbarker) [reply libraries/AP_Button/AP_Button.cpp]: Redone in terms of finalizing a `const char*` chunk of data on each channel.

**Comment by peterbarker** [comment]
Need to test mavlink1 vs mavlink2 interactions on this PR.

**Comment by tridge** [comment]
I have tested this for correct handling of mavlink1/2 and it works nicely

**Pull request merged**